go/vt/vttablet/grpctmclient/cached_client.go

```diff
@@ -0,0 +1,340 @@
+	DialTimings          *stats.Timings
+	EvictionQueueTimings *stats.Timings
+}{
+	ConnReuse:    stats.NewGauge("tabletmanagerclient_cachedconn_reuse", "number of times a call to dial() was able to reuse an existing connection"),
+	ConnNew:      stats.NewGauge("tabletmanagerclient_cachedconn_new", "number of times a call to dial() resulted in a dialing a new grpc clientconn"),
+	DialTimeouts: stats.NewGauge("tabletmanagerclient_cachedconn_dial_timeouts", "number of context timeouts during dial()"),
+	DialTimings:  stats.NewTimings("tabletmanagerclient_cachedconn_dialtimings", "timings for various dial paths", "path", "rlock_fast", "sema_fast", "sema_poll"),
+	// TODO: add timings for heap operations (push, pop, fix)
+}
+
+// NewCachedConnClient returns a grpc Client using the priority queue cache
+// dialer implementation.
+func NewCachedConnClient(capacity int) *Client {
+	dialer := &cachedConnDialer{
+		conns:        make(map[string]*cachedConn, capacity),
+		queue:        make(cachedConns, 0, capacity),
+		connWaitSema: sync2.NewSemaphore(capacity, 0),
+	}
+
+	heap.Init(&dialer.queue)
+	return &Client{dialer}
+}
+
+var _ dialer = (*cachedConnDialer)(nil)
+
+func (dialer *cachedConnDialer) dial(ctx context.Context, tablet *topodatapb.Tablet) (tabletmanagerservicepb.TabletManagerClient, io.Closer, error) {
+	start := time.Now()
+
+	addr := getTabletAddr(tablet)
+	dialer.m.RLock()
+	if conn, ok := dialer.conns[addr]; ok {
+		defer func() {
+			dialerStats.DialTimings.Add("rlock_fast", time.Since(start))
+		}()
+		defer dialer.m.RUnlock()
+		return dialer.redial(conn)
+	}
+	dialer.m.RUnlock()
+
+	if dialer.connWaitSema.TryAcquire() {
+		defer func() {
+			dialerStats.DialTimings.Add("sema_fast", time.Since(start))
+		}()
+		dialer.m.Lock()
+		defer dialer.m.Unlock()
+
+		// Check if another goroutine managed to dial a conn for the same addr
+		// while we were waiting for the write lock. This is identical to the
+		// read-lock section above.
+		if conn, ok := dialer.conns[addr]; ok {
+			return dialer.redial(conn)
+		}
+
+		return dialer.newdial(addr, true /* manage queue lock */)
+	}
+
+	defer func() {
+		dialerStats.DialTimings.Add("sema_poll", time.Since(start))
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			dialerStats.DialTimeouts.Add(1)
+			return nil, nil, ctx.Err()
+		default:
+			dialer.m.Lock()
+			if conn, ok := dialer.conns[addr]; ok {
+				// Someone else dialed this addr while we were polling. No need
+				// to evict anyone else, just reuse the existing conn.
+				defer dialer.m.Unlock()
+				return dialer.redial(conn)
+			}
+
+			dialer.qMu.Lock()
+			conn := dialer.queue[0]
+			if conn.refs != 0 {
+				dialer.qMu.Unlock()
+				dialer.m.Unlock()
+				continue
+			}
+
+			// We're going to return from this point
+			defer dialer.m.Unlock()
+			defer dialer.qMu.Unlock()
+			heap.Pop(&dialer.queue)
+			delete(dialer.conns, conn.key)
+			conn.cc.Close()
+
+			return dialer.newdial(addr, false /* manage queue lock */)
```

Comment:
reading the contract with newdial below, we don't have the connWaitSema acquired (I don't think), and try to release it in newdial.

Reply:
great catch, thank you! i'm going to make the decomposition refactors you suggested and if see if a cleaner way to handle this falls out of that, otherwise i'll just add another bool arg to newdial

Reply:
okay this one took me a while, but i realized during the decomposition that this was not a mistake, it's just not very clear.

semantically, we want:
* `Acquire()` anytime succeed at `newdial`, because we are taking a slot in the cache
* `Release()` anytime we evict, because we are freeing a slot in the cache.

in this code path, we have evicted, and so semantically, we should jump back up to the `TryAcquire`, and enter the `newdial` path that way (where the `Release` contract is much clearer). however, we're giving priority to the `dial` call that evicted a conn by giving them the chance to grab the cache slot they just freed (which is why we don't `Release` when we `Pop` from the heap). so, if the `newdial` succeeds, we've skipped the `Release`+`Acquire` calls, and the sema is left in the same state because we've traded one evicted conn for one new conn. if however, `newdial` fails, then we `Release` (which logically _should_ have happened as soon as we `Pop`-ped), and leave the sema with one slot free for the one conn that we evicted but did not replace.